UnifiedPDF: Support for loading PDFs with #page= fragments

hortont424 opened the pull request:
#### 63dde78713856ee9925a6802d16d971d1547babf
<pre>
UnifiedPDF: Support for loading PDFs with #page= fragments
<a href="https://bugs.webkit.org/show_bug.cgi?id=269180">https://bugs.webkit.org/show_bug.cgi?id=269180</a>
<a href="https://rdar.apple.com/122765668">rdar://122765668</a>

Reviewed by Simon Fraser.

* Source/WebCore/plugins/PluginViewBase.h:
(WebCore::PluginViewBase::didAttachScrollingNode):
* Source/WebCore/rendering/RenderEmbeddedObject.cpp:
(WebCore::RenderEmbeddedObject::didAttachScrollingNode):
* Source/WebCore/rendering/RenderEmbeddedObject.h:
* Source/WebCore/rendering/RenderLayerCompositor.cpp:
(WebCore::RenderLayerCompositor::attachWidgetContentLayers):
* Source/WebKit/WebProcess/Plugins/PDF/PDFPluginBase.h:
(WebKit::PDFPluginBase::didAttachScrollingNode):
* Source/WebKit/WebProcess/Plugins/PluginView.cpp:
(WebKit::PluginView::didAttachScrollingNode):
* Source/WebKit/WebProcess/Plugins/PluginView.h:
Let the plugin know when it&apos;s in the scrolling tree and it&apos;s safe to start scrolling.

* Source/WebKit/WebProcess/Plugins/PDF/UnifiedPDF/UnifiedPDFPlugin.h:
* Source/WebKit/WebProcess/Plugins/PDF/UnifiedPDF/UnifiedPDFPlugin.mm:
(WebKit::UnifiedPDFPlugin::installPDFDocument):
(WebKit::UnifiedPDFPlugin::didAttachScrollingNode):
Scroll to the fragment in the URL whenever *both* the PDF document is installed and we&apos;re in the scrolling tree.

(WebKit::UnifiedPDFPlugin::pageHeight const):
(WebKit::UnifiedPDFPlugin::firstPageHeight const):
Factor out pageHeight() from firstPageHeight().

(WebKit::UnifiedPDFPlugin::scrollToPDFDestination):
Fix a mistake in here: indexForPage is zero-based like PageIndex, but the origin
is flipped, so the y origin of the page is at its height instead of 0 (which is
how I was misled into subtracting 1 from the page index...).

(WebKit::UnifiedPDFPlugin::scrollToPointInPage):
Fix the local name here to say `contents` instead of `document`.
Rename this to &quot;point in page&quot;, to go along with...

(WebKit::UnifiedPDFPlugin::scrollToPage):
... this new function, which just scrolls to the origin of a page.

(WebKit::UnifiedPDFPlugin::scrollToFragmentIfNeeded):
Parse the fragment out. For now, only support page=; for parity with
PDFPlugin we also need to support nameddest=.

* Source/WebKit/WebProcess/Plugins/PDF/PDFPluginBase.h:
(WebKit::PDFPluginBase::didSameDocumentNavigationForFrame):
* Source/WebKit/WebProcess/Plugins/PluginView.cpp:
(WebKit::PluginView::didSameDocumentNavigationForFrame):
* Source/WebKit/WebProcess/Plugins/PluginView.h:
* Source/WebKit/WebProcess/WebPage/WebPage.cpp:
(WebKit::WebPage::didSameDocumentNavigationForFrame):
Plumb fragment changes through to the plugin, so that dynamically editing
the URL navigates to the new fragment (instead of having to force a reload).

Canonical link: <a href="https://commits.webkit.org/274496@main">https://commits.webkit.org/274496@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/54329afb37a86a4e8bca49e853be0f847b97345d

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/39261 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/18240 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/41614 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/41795 "Built successfully") | [  ~~🛠 wincairo~~](https://ews-build.webkit.org/#/builders/32/builds/35161 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/41567 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/21097 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/15569 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/32851 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/39835 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/15358 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/34031 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/13345 "Passed tests") | 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/42/builds/13311 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/34971 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/43073 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/35653 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/35300 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/39117 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [  ~~🛠 tv~~](https://ews-build.webkit.org/#/builders/44/builds/14073 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/50/builds/11615 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/37356 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/15679 "Built successfully") | | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/8785 "Built successfully and passed tests") | [  ~~🛠 watch~~](https://ews-build.webkit.org/#/builders/43/builds/15342 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| | [  ~~🛠 watch-sim~~](https://ews-build.webkit.org/#/builders/45/builds/15165 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
<!--EWS-Status-Bubble-End-->

hortont424 commented on Source/WebKit/WebProcess/Plugins/PDF/UnifiedPDF/UnifiedPDFPlugin.mm:
```diff
@@ -1354,19 +1368,72 @@ static bool isContextMenuEvent(const WebMouseEvent& event)
+    };
+
+    if (auto remainder = remainderForPrefix("page="_s)) {
+        WTFLogAlways("page %s", remainder->utf8().data());
```
Did not mean to leave this in.

smfr commented on Source/WebKit/WebProcess/Plugins/PDF/UnifiedPDF/UnifiedPDFPlugin.h:
```diff
@@ -137,6 +137,8 @@ class UnifiedPDFPlugin final : public PDFPluginBase, public WebCore::GraphicsLay
     WebCore::IntSize documentSize() const;
     WebCore::IntSize contentsSize() const override;
     unsigned firstPageHeight() const override;
+    unsigned pageHeight(PDFDocumentLayout::PageIndex) const;
```
I like to call things like this `heightForPage()` which sounds less like something that would apply to all pages.

smfr commented on Source/WebKit/WebProcess/Plugins/PDF/UnifiedPDF/UnifiedPDFPlugin.mm:
```diff
@@ -719,11 +728,16 @@ static String mutationObserverNotificationString()
     if (isLocked() || !m_documentLayout.pageCount())
         return 0;
-    return static_cast<unsigned>(CGCeiling(m_documentLayout.boundsForPageAtIndex(0).height()));
+    return static_cast<unsigned>(CGCeiling(m_documentLayout.boundsForPageAtIndex(pageIndex).height()));
```
It's a bit weird to see `CGCeiling` here; maybe `std::ceil()`?

smfr commented on Source/WebKit/WebProcess/Plugins/PDF/UnifiedPDF/UnifiedPDFPlugin.mm:
```diff
@@ -1354,19 +1368,72 @@ static bool isContextMenuEvent(const WebMouseEvent& event)
+
+void UnifiedPDFPlugin::scrollToPage(PDFDocumentLayout::PageIndex pageIndex)
+{
+    scrollToPointInPage({ 0, static_cast<int>(pageHeight(pageIndex)) }, pageIndex);
```
Is the `pageHeight()` because of flipped coordinates in the page? Is a bit mysterious; a comment or some `topOfPage()` helper  would clarify.

smfr commented on Source/WebKit/WebProcess/Plugins/PDF/UnifiedPDF/UnifiedPDFPlugin.mm:
```diff
@@ -1354,19 +1368,72 @@ static bool isContextMenuEvent(const WebMouseEvent& event)
+    auto fragmentView = StringView(fragment);
+
+    // Only respect the first fragment component.
+    if (auto endOfFirstComponentLocation = fragmentView.find('&'); endOfFirstComponentLocation != notFound)
```
Are there no helpers for these? Or are they PDF specific?

hortont424 replied:
We have query string parsers, this is like query format in the fragment. It's derived indirectly from the PDF spec which I guess is a bit more restrictive than a naive impl would be 🤷‍♂️

smfr commented on Source/WebKit/WebProcess/Plugins/PDF/UnifiedPDF/UnifiedPDFPlugin.mm:
```diff
@@ -1354,19 +1368,72 @@ static bool isContextMenuEvent(const WebMouseEvent& event)
+        return;
+    }
+
+    // FIXME: Support named destinations.
```
Add a bug number? Not clear how important this is.

hortont424 committed:
UnifiedPDF: Support for loading PDFs with #page= fragments
https://bugs.webkit.org/show_bug.cgi?id=269180
rdar://122765668

Reviewed by Simon Fraser.

* Source/WebCore/plugins/PluginViewBase.h:
(WebCore::PluginViewBase::didAttachScrollingNode):
* Source/WebCore/rendering/RenderEmbeddedObject.cpp:
(WebCore::RenderEmbeddedObject::didAttachScrollingNode):
* Source/WebCore/rendering/RenderEmbeddedObject.h:
* Source/WebCore/rendering/RenderLayerCompositor.cpp:
(WebCore::RenderLayerCompositor::attachWidgetContentLayers):
* Source/WebKit/WebProcess/Plugins/PDF/PDFPluginBase.h:
(WebKit::PDFPluginBase::didAttachScrollingNode):
* Source/WebKit/WebProcess/Plugins/PluginView.cpp:
(WebKit::PluginView::didAttachScrollingNode):
* Source/WebKit/WebProcess/Plugins/PluginView.h:
Let the plugin know when it's in the scrolling tree and it's safe to start scrolling.

* Source/WebKit/WebProcess/Plugins/PDF/UnifiedPDF/UnifiedPDFPlugin.h:
* Source/WebKit/WebProcess/Plugins/PDF/UnifiedPDF/UnifiedPDFPlugin.mm:
(WebKit::UnifiedPDFPlugin::installPDFDocument):
(WebKit::UnifiedPDFPlugin::didAttachScrollingNode):
Scroll to the fragment in the URL whenever *both* the PDF document is installed and we're in the scrolling tree.

(WebKit::UnifiedPDFPlugin::pageHeight const):
(WebKit::UnifiedPDFPlugin::firstPageHeight const):
Factor out pageHeight() from firstPageHeight().

(WebKit::UnifiedPDFPlugin::scrollToPDFDestination):
Fix a mistake in here: indexForPage is zero-based like PageIndex, but the origin
is flipped, so the y origin of the page is at its height instead of 0 (which is
how I was misled into subtracting 1 from the page index...).

(WebKit::UnifiedPDFPlugin::scrollToPointInPage):
Fix the local name here to say `contents` instead of `document`.
Rename this to "point in page", to go along with...

(WebKit::UnifiedPDFPlugin::scrollToPage):
... this new function, which just scrolls to the origin of a page.

(WebKit::UnifiedPDFPlugin::scrollToFragmentIfNeeded):
Parse the fragment out. For now, only support page=; for parity with
PDFPlugin we also need to support nameddest=.

* Source/WebKit/WebProcess/Plugins/PDF/PDFPluginBase.h:
(WebKit::PDFPluginBase::didSameDocumentNavigationForFrame):
* Source/WebKit/WebProcess/Plugins/PluginView.cpp:
(WebKit::PluginView::didSameDocumentNavigationForFrame):
* Source/WebKit/WebProcess/Plugins/PluginView.h:
* Source/WebKit/WebProcess/WebPage/WebPage.cpp:
(WebKit::WebPage::didSameDocumentNavigationForFrame):
Plumb fragment changes through to the plugin, so that dynamically editing
the URL navigates to the new fragment (instead of having to force a reload).

Canonical link: https://commits.webkit.org/274496@main
